Update server examples with scopes and collections - N1QL DML pages (#1809)
couchbase/docs-server: modules/n1ql/pages/n1ql-language-reference/upsert.adoc

For more details about user roles, see
{authorization-overview}[Authorization].

> WARNING: Please note that the examples below will alter the data in your sample buckets.

Reviewer: How to mitigate
```suggestion
WARNING: Please note that the examples below will alter the data in your sample buckets.
To restore your sample data, remove and reinstall the `travel-sample` bucket.
Refer to xref:manage:manage-settings/install-sample-buckets.adoc[Sample Buckets] for details. 
```

UPSERT INTO `travel-sample`.inventory.hotel (KEY, VALUE)
VALUES ("key1", { "type" : "hotel", "name" : "new hotel" });
----

* To execute the following statement, user must have the _Query Update_ and _Query Insert_ privileges on the `pass:c[`travel-sample`]` bucket and _Query Select_ privilege on `pass:c[`travel-sample`]` (for RETURNING clause).
+
[source,n1ql]
----
UPSERT INTO `travel-sample` (KEY, VALUE)
UPSERT INTO `travel-sample`.inventory.hotel (KEY, VALUE)
VALUES ("key1", { "type" : "hotel", "name" : "new hotel" })
RETURNING *
RETURNING *;
----
+
.Result:
[source,json]
----
[
  {
    "hotel": {
      "name": "new hotel",
      "type": "hotel"
    }
  }
]
----

* To execute the following statement, user must have the _Query Update_ and _Query Insert_ privileges on the `pass:c[`travel-sample`]` bucket and _Query Select_ privilege on `pass:c[`beer-sample`]`.
* To execute the following statement, you must have the _Query Update_ and _Query Insert_ privileges on the `pass:c[`travel-sample`.inventory.hotel]` bucket and _Query Select_ privilege on `pass:c[`beer-sample`]`.
+
[source,n1ql]
----
UPSERT INTO `travel-sample` (KEY foo, VALUE bar)

> SELECT foo, bar FROM `beer-sample`

Author: Not completely sure what we were trying to showcase here previously, this query doesn't seem to work. Any ideas?

The error returned is:
```
 "Cannot UPSERT non-string key Missing field or index foo. of type value.missingValue."
```

Reviewer: Yes this query wouldn't have worked previously. It's trying to insert records (key-value pairs) where the key is defined by the alias foo and the value is defined by the alias bar. In order for the insert to work, we have to set foo to the id of an existing record and set bar to the content of an existing record. See suggestion below.

> UPSERT INTO `travel-sample`.inventory.hotel (KEY foo, VALUE bar)
> SELECT foo, bar FROM `beer-sample`;

Reviewer: Using landmark because a brewery is more of a landmark than a hotel
```suggestion
UPSERT INTO `travel-sample`.inventory.landmark (KEY foo, VALUE bar)
SELECT META(doc).id AS foo, doc AS bar FROM `beer-sample` AS doc WHERE type = "brewery";
```

For more details about user roles, see
{authorization-overview}[Authorization].

WARNING: Please note that the examples below will alter the data in your sample buckets.
To restore your sample data, remove and reinstall the `travel-sample` bucket.
Refer to xref:manage:manage-settings/install-sample-buckets.adoc[Sample Buckets] for details.

For example,

> * To execute the following statement, user must have the _Query Update_ and _Query Insert_ privileges on `pass:c[`travel-sample`]`.

Reviewer: ```suggestion
* To execute the following statement, you must have the _Query Update_ and _Query Insert_ privileges on `pass:c[`travel-sample`].inventory.hotel`.
```

+
[source,n1ql]
----
UPSERT INTO `travel-sample`.inventory.hotel (KEY, VALUE)
VALUES ("key1", { "type" : "hotel", "name" : "new hotel" });
----

> * To execute the following statement, user must have the _Query Update_ and _Query Insert_ privileges on the `pass:c[`travel-sample`]` bucket and _Query Select_ privilege on `pass:c[`travel-sample`]` (for RETURNING clause).

Reviewer: ```suggestion
* To execute the following statement, you must have the _Query Update_ and _Query Insert_ privileges on `pass:c[`travel-sample`].inventory.hotel` and the _Query Select_ privilege on `pass:c[`travel-sample`].inventory.hotel` also (for RETURNING clause).
```

+
[source,n1ql]
----
UPSERT INTO `travel-sample`.inventory.hotel (KEY, VALUE)
VALUES ("key1", { "type" : "hotel", "name" : "new hotel" })
RETURNING *;
----
+
.Result:
[source,json]
----
[
  {
    "hotel": {
      "name": "new hotel",
      "type": "hotel"
    }
  }
]
----

> * To execute the following statement, you must have the _Query Update_ and _Query Insert_ privileges on the `pass:c[`travel-sample`.inventory.landmark]` bucket and _Query Select_ privilege on `pass:c[`beer-sample`]`.

Reviewer: It's a collection, not a bucket :smiley:
```suggestion
* To execute the following statement, you must have the _Query Update_ and _Query Insert_ privileges on `pass:c[`travel-sample`.inventory.landmark]` and _Query Select_ privilege on `pass:c[`beer-sample`]`.
```

> .Query

Reviewer: Update the description for example-1 (line 218):
```
The following statement upserts documents with type [.in]`landmark-pub` into the [.in]`landmark` keyspace.
```
